File: platform/polyglot/confluence_api/mjs/src/services/webhook-service.mjs
/**
 * @module services/webhook-service
 * @description Service for Confluence Webhook REST API operations.
 *
 * Manages webhooks (also called "listeners") for receiving event notifications
 * from Confluence. Supports CRUD, testing, and invocation/statistics retrieval.
 *
 * Confluence Data Center REST API v9.2.3.
 */

import { createLogger } from '../logger.mjs';

const log = createLogger('confluence_api', import.meta.url);

/**
 * Service class for Confluence webhook operations.
 */
export class WebhookService {
  /**
   * @param {import('../client/FetchClient.mjs').FetchClient} client
   *   The Confluence fetch client instance.
   */
  constructor(client) {
    /** @private */
    this._client = client;
  }

  /**
   * Get all registered webhooks.
   *
   * @returns {Promise<Object>} List of webhooks.
   */
  async getWebhooks() {
    log.debug('getWebhooks called');
    try {
      const result = await this._client.get('webhook');
      log.info('getWebhooks succeeded', { count: result.results?.length ?? result.length });
      return result;
    } catch (error) {
      log.error('getWebhooks failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Create a new webhook.
   *
   * @param {Object} data - Webhook creation payload.
   * @param {string} data.url - The callback URL.
   * @param {string[]} data.events - Event types to listen for.
   * @param {string} [data.name] - Webhook name.
   * @param {boolean} [data.active] - Whether the webhook is active.
   * @returns {Promise<Object>} The created webhook object.
   */
  async createWebhook(data) {
    log.debug('createWebhook called', { url: data?.url, events: data?.events });
    try {
      const result = await this._client.post('webhook', data);
      log.info('createWebhook succeeded', { id: result.id, url: result.url });
      return result;
    } catch (error) {
      log.error('createWebhook failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get a single webhook by ID.
   *
   * @param {string} webhookId - The webhook ID.
   * @returns {Promise<Object>} The webhook object.
   */
  async getWebhook(webhookId) {
    log.debug('getWebhook called', { webhookId });
    try {
      const result = await this._client.get(`webhook/${webhookId}`);
      log.info('getWebhook succeeded', { webhookId, url: result.url });
      return result;
    } catch (error) {
      log.error('getWebhook failed', { webhookId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Update an existing webhook.
   *
   * @param {string} webhookId - The webhook ID.
   * @param {Object} data - Webhook update payload.
   * @returns {Promise<Object>} The updated webhook object.
   */
  async updateWebhook(webhookId, data) {
    log.debug('updateWebhook called', { webhookId });
    try {
      const result = await this._client.put(`webhook/${webhookId}`, data);
      log.info('updateWebhook succeeded', { webhookId });
      return result;
    } catch (error) {
      log.error('updateWebhook failed', { webhookId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Delete a webhook.
   *
   * @param {string} webhookId - The webhook ID.
   * @returns {Promise<void>}
   */
  async deleteWebhook(webhookId) {
    log.debug('deleteWebhook called', { webhookId });
    try {
      const result = await this._client.delete(`webhook/${webhookId}`);
      log.info('deleteWebhook succeeded', { webhookId });
      return result;
    } catch (error) {
      log.error('deleteWebhook failed', { webhookId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Send a test event to a webhook endpoint.
   *
   * @param {Object} data - Test payload containing the webhook URL and event type.
   * @returns {Promise<Object>} Test result.
   */
  async testWebhook(data) {
    log.debug('testWebhook called', { url: data?.url });
    try {
      const result = await this._client.post('webhook/test', data);
      log.info('testWebhook succeeded');
      return result;
    } catch (error) {
      log.error('testWebhook failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get the latest invocations (delivery history) for a webhook.
   *
   * @param {string} webhookId - The webhook ID.
   * @returns {Promise<Object>} Latest invocation records.
   */
  async getLatestInvocations(webhookId) {
    log.debug('getLatestInvocations called', { webhookId });
    try {
      const result = await this._client.get(`webhook/${webhookId}/latest`);
      log.info('getLatestInvocations succeeded', { webhookId });
      return result;
    } catch (error) {
      log.error('getLatestInvocations failed', { webhookId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get statistics for a specific webhook.
   *
   * @param {string} webhookId - The webhook ID.
   * @returns {Promise<Object>} Webhook statistics.
   */
  async getWebhookStatistics(webhookId) {
    log.debug('getWebhookStatistics called', { webhookId });
    try {
      const result = await this._client.get(`webhook/${webhookId}/statistics`);
      log.info('getWebhookStatistics succeeded', { webhookId });
      return result;
    } catch (error) {
      log.error('getWebhookStatistics failed', { webhookId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get a summary of statistics for a webhook.
   *
   * @param {string} webhookId - The webhook ID.
   * @returns {Promise<Object>} Webhook statistics summary.
   */
  async getWebhookStatisticsSummary(webhookId) {
    log.debug('getWebhookStatisticsSummary called', { webhookId });
    try {
      const result = await this._client.get(`webhook/${webhookId}/statistics/summary`);
      log.info('getWebhookStatisticsSummary succeeded', { webhookId });
      return result;
    } catch (error) {
      log.error('getWebhookStatisticsSummary failed', { webhookId, message: error.message, status: error.status });
      throw error;
    }
  }
}

File: platform/polyglot/confluence_api/py/src/confluence_api/services/webhook_service.py
"""
Webhook service for Confluence Data Center REST API v9.2.3.

Provides CRUD operations for webhooks including testing,
invocation history, and statistics.
"""

from __future__ import annotations

from typing import Any

from confluence_api.logger import create_logger

log = create_logger("confluence_api", __file__)


class WebhookService:
    """Service for Confluence webhook operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    def get_webhooks(self) -> dict:
        """Retrieve all registered webhooks."""
        log.debug("get_webhooks called")
        result = self._client.get("webhook")
        log.info("get_webhooks succeeded")
        return result

    def create_webhook(self, data: dict) -> dict:
        """
        Register a new webhook.

        Args:
            data: Webhook configuration including url, events, etc.

        Returns:
            Created webhook data dict.
        """
        log.debug("create_webhook called", {"url": data.get("url")})
        result = self._client.post("webhook", json_data=data)
        log.info("create_webhook succeeded", {"id": result.get("id")})
        return result

    def get_webhook(self, webhook_id: int) -> dict:
        """Retrieve a specific webhook by ID."""
        log.debug("get_webhook called", {"webhook_id": webhook_id})
        result = self._client.get(f"webhook/{webhook_id}")
        log.info("get_webhook succeeded", {"webhook_id": webhook_id})
        return result

    def update_webhook(self, webhook_id: int, data: dict) -> dict:
        """Update an existing webhook."""
        log.debug("update_webhook called", {"webhook_id": webhook_id})
        result = self._client.put(f"webhook/{webhook_id}", json_data=data)
        log.info("update_webhook succeeded", {"webhook_id": webhook_id})
        return result

    def delete_webhook(self, webhook_id: int) -> dict:
        """Delete a webhook by ID."""
        log.debug("delete_webhook called", {"webhook_id": webhook_id})
        result = self._client.delete(f"webhook/{webhook_id}")
        log.info("delete_webhook succeeded", {"webhook_id": webhook_id})
        return result

    def test_webhook(self, data: dict) -> dict:
        """
        Send a test event to a webhook URL.

        Args:
            data: Test payload configuration.

        Returns:
            Test result dict.
        """
        log.debug("test_webhook called", {"url": data.get("url")})
        result = self._client.post("webhook/test", json_data=data)
        log.info("test_webhook succeeded")
        return result

    def get_latest_invocations(self, webhook_id: int) -> dict:
        """Retrieve the latest invocation history for a webhook."""
        log.debug("get_latest_invocations called", {"webhook_id": webhook_id})
        result = self._client.get(f"webhook/{webhook_id}/history")
        log.info("get_latest_invocations succeeded", {"webhook_id": webhook_id})
        return result

    def get_webhook_statistics(self, webhook_id: int) -> dict:
        """Retrieve detailed statistics for a specific webhook."""
        log.debug("get_webhook_statistics called", {"webhook_id": webhook_id})
        result = self._client.get(f"webhook/{webhook_id}/statistics")
        log.info("get_webhook_statistics succeeded", {"webhook_id": webhook_id})
        return result

    def get_webhook_statistics_summary(self, webhook_id: int) -> dict:
        """Retrieve a summary of statistics for a specific webhook."""
        log.debug("get_webhook_statistics_summary called", {"webhook_id": webhook_id})
        result = self._client.get(f"webhook/{webhook_id}/statistics/summary")
        log.info("get_webhook_statistics_summary succeeded", {"webhook_id": webhook_id})
        return result


class AsyncWebhookService:
    """Async service for Confluence webhook operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    async def get_webhooks(self) -> dict:
        """Retrieve all registered webhooks."""
        log.debug("get_webhooks called")
        result = await self._client.get("webhook")
        log.info("get_webhooks succeeded")
        return result

    async def create_webhook(self, data: dict) -> dict:
        """Register a new webhook."""
        log.debug("create_webhook called", {"url": data.get("url")})
        result = await self._client.post("webhook", json_data=data)
        log.info("create_webhook succeeded", {"id": result.get("id")})
        return result

    async def get_webhook(self, webhook_id: int) -> dict:
        """Retrieve a specific webhook by ID."""
        log.debug("get_webhook called", {"webhook_id": webhook_id})
        result = await self._client.get(f"webhook/{webhook_id}")
        log.info("get_webhook succeeded", {"webhook_id": webhook_id})
        return result

    async def update_webhook(self, webhook_id: int, data: dict) -> dict:
        """Update an existing webhook."""
        log.debug("update_webhook called", {"webhook_id": webhook_id})
        result = await self._client.put(f"webhook/{webhook_id}", json_data=data)
        log.info("update_webhook succeeded", {"webhook_id": webhook_id})
        return result

    async def delete_webhook(self, webhook_id: int) -> dict:
        """Delete a webhook by ID."""
        log.debug("delete_webhook called", {"webhook_id": webhook_id})
        result = await self._client.delete(f"webhook/{webhook_id}")
        log.info("delete_webhook succeeded", {"webhook_id": webhook_id})
        return result

    async def test_webhook(self, data: dict) -> dict:
        """Send a test event to a webhook URL."""
        log.debug("test_webhook called", {"url": data.get("url")})
        result = await self._client.post("webhook/test", json_data=data)
        log.info("test_webhook succeeded")
        return result

    async def get_latest_invocations(self, webhook_id: int) -> dict:
        """Retrieve the latest invocation history for a webhook."""
        log.debug("get_latest_invocations called", {"webhook_id": webhook_id})
        result = await self._client.get(f"webhook/{webhook_id}/history")
        log.info("get_latest_invocations succeeded", {"webhook_id": webhook_id})
        return result

    async def get_webhook_statistics(self, webhook_id: int) -> dict:
        """Retrieve detailed statistics for a specific webhook."""
        log.debug("get_webhook_statistics called", {"webhook_id": webhook_id})
        result = await self._client.get(f"webhook/{webhook_id}/statistics")
        log.info("get_webhook_statistics succeeded", {"webhook_id": webhook_id})
        return result

    async def get_webhook_statistics_summary(self, webhook_id: int) -> dict:
        """Retrieve a summary of statistics for a specific webhook."""
        log.debug("get_webhook_statistics_summary called", {"webhook_id": webhook_id})
        result = await self._client.get(f"webhook/{webhook_id}/statistics/summary")
        log.info("get_webhook_statistics_summary succeeded", {"webhook_id": webhook_id})
        return result
